experiments/composite/pkg/watchset/dependencies.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchset
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+)
+
+type resourceVersion int64
+
+type DependencySet struct {
+	// Note: GroupVersionResource (instead of GroupResource) because we typically will only care about one version anyway,
+	// and because sometimes we do care - when we want to read the object.
+	objects map[schema.GroupVersionResource]map[types.NamespacedName]resourceVersion
+
+	lists map[schema.GroupVersionResource]map[listOpKey]listOp
+}
+
+func newDependencySet() *DependencySet {
+	return &DependencySet{
+		objects: make(map[schema.GroupVersionResource]map[types.NamespacedName]resourceVersion),
+		lists:   make(map[schema.GroupVersionResource]map[listOpKey]listOp),
+	}
+}
+
+var _ fmt.Stringer = dependencySet{}
+
+func (x *DependencySet) String() string {
+	var sb strings.Builder
+	for gvk, objects := range x.objects {
+		if sb.Len() != 0 {
+			fmt.Fprintf(&sb, ",")
```

Comment:
Why Fprintf rather than sb.WriteString?

Reply:
So in other places we do this:

`fmt.Fprintf(&sb, "%v@%d", nn.String(), rv)`

Sadly there's some sb.WriteStringF, so we would have to do `sb.WriteString(fmt.Sprintf("%v@%d", nn.String(), rv))`

But then this was trying to be consistent and not alternate between sb.WriteString and fmt.Fprintf.

I'm honestly not sure which is better... I can see a case for any of mixing or WriteString everywhere or Fprintf everywhere.

Reply:
That makes sense. Given the API you have to work with.